### PR TITLE
8257848: -XX:CompileCommand=blackhole,* should be diagnostic

### PR DESCRIPTION
**Pull request by shipilev**
This follows up on retroactive CSR review comments (JDK-8257827). As the result of discussion, it is decided to demote the "blackhole" compile command to diagnostic. Examples:

```
$ build/linux-x86_64-server-release/images/jdk/bin/java -Xcomp -XX:CompileCommand=quiet -XX:CompileCommand=blackhole,java/lang/Object.toString
OpenJDK 64-Bit Server VM warning: Blackhole compile option is diagnostic and must be enabled via -XX:+UnlockDiagnosticVMOptions

$ build/linux-x86_64-server-release/images/jdk/bin/java -Xcomp -XX:+UnlockDiagnosticVMOptions -XX:CompileCommand=quiet -XX:CompileCommand=blackhole,java/lang/Object.toString
OpenJDK 64-Bit Server VM warning: Blackhole compile option only works for methods with void type: java.lang.Object.toString()Ljava/lang/String;
```

New test verifies the locking/unlocking of the command.

Testing:
 - [x] ad-hoc `compiler/blackhole` test
 - [x] validation runs with JMH
 - [x] runs in `tier1` (needs #1654 to be merged)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257848](https://bugs.openjdk.java.net/browse/JDK-8257848): -XX:CompileCommand=blackhole,* should be diagnostic

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to cbc07869e3fd2455a9504f0478774ab0cd56ea92

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1674/head:pull/1674`
`$ git checkout pull/1674`

### CODE REVIEW

File: src/hotspot/share/compiler/compilerOracle.cpp [hunk 1]
```diff
@@ -410,15 +410,19 @@ bool CompilerOracle::should_break_at(const methodHandle& method) {
...
 bool CompilerOracle::should_blackhole(const methodHandle& method) {
-  if (check_predicate(CompileCommand::Blackhole, method)) {
-    if (method->result_type() == T_VOID) {
-      return true;
-    } else {
-      warning("blackhole compile command only works for methods with void type: %s",
-              method->name_and_sig_as_C_string());
-    }
+  if (!check_predicate(CompileCommand::Blackhole, method)) {
+    return false;
```
Comment (shipilev) [inline src/hotspot/share/compiler/compilerOracle.cpp:414]: Review self-comment. In fact, I am not sure if this should be `check_predicate` (like other commands), or `has_option` (like other options). CompilerOracle refactoring in JDK-8256508 seem to have converged both to mean the same. I left `check_predicate` as in original, and it also matches other `should_*`, see the definitions just above.

File: src/hotspot/share/compiler/compilerOracle.cpp [hunk 2]
```diff
@@ -410,15 +410,19 @@ bool CompilerOracle::should_break_at(const methodHandle& method) {
...
-      return true;
-    } else {
-      warning("blackhole compile command only works for methods with void type: %s",
-              method->name_and_sig_as_C_string());
-    }
+  if (!has_option(method, CompileCommand::Blackhole)) {
+    return false;
   }
-  return false;
+  if (!UnlockDiagnosticVMOptions) {
```
Comment (iwanowww) [inline src/hotspot/share/compiler/compilerOracle.cpp:416]: The check happens too late (during compilation when JIT encounters marked method).

IMO it should happen during command-line parsing at startup: `CompilerOracle::parse_from_line()` or `register_command()` look suitable.

**Review by iwanowww — CHANGES_REQUESTED** [review]

File: src/hotspot/share/compiler/compilerOracle.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/compiler/compilerOracle.cpp:414]: I think using `check_predicate` is perfectly fine in this case.

Comment (shipilev) [reply src/hotspot/share/compiler/compilerOracle.cpp:414]: Okay, I shall leave it at that then.

File: src/hotspot/share/compiler/compilerOracle.cpp (hunk 2 above)
Comment (shipilev) [reply src/hotspot/share/compiler/compilerOracle.cpp]: Right. I moved it to `register_command`, and `guarantee`-d in `CompilerOracle::should_blackhole` that we ever get there (that is, predicate matched) if diagnostic options were unlocked. Still passes tests in both release and fastdebug.

**Review by vnkozlov — APPROVED** [review]

File: src/hotspot/share/compiler/compilerOracle.cpp [hunk 3]
```diff
@@ -410,15 +415,16 @@ bool CompilerOracle::should_break_at(const methodHandle& method) {
...
+  if (!check_predicate(CompileCommand::Blackhole, method)) {
+    return false;
   }
-  return false;
+  if (method->result_type() != T_VOID) {
+    warning("Blackhole compile option only works for methods with void type: %s",
+            method->name_and_sig_as_C_string());
+    return false;
+  }
+  guarantee(UnlockDiagnosticVMOptions, "Checked during initial parsing");
```
Comment (iwanowww) [inline src/hotspot/share/compiler/compilerOracle.cpp:426]: You can move it up right after `check_predicate`: `check_predicate(CompileCommand::Blackhole, method)` is allowed to return `true` iff `UnlockDiagnosticVMOptions` is set. 

Not sure `guarantee` is well-justified here, IMO `assert` would suffice.

**Review by iwanowww — APPROVED** [review]

**Review by vnkozlov — COMMENTED** [review]
Do you need to update blackhole tests to add -XX:+UnlockDiagnosticVMOptions?

File: src/hotspot/share/compiler/compilerOracle.cpp (hunk 3 above)
Comment (shipilev) [reply src/hotspot/share/compiler/compilerOracle.cpp:426]: `UnlockDiagnosticVMOptions` is `trueInDebug`, so `assert` is useless. We really care about "release" mode here.

**Comment by shipilev** [comment]
> Do you need to update blackhole tests to add -XX:+UnlockDiagnosticVMOptions?

They already have it, because they use `-XX:+AbortVMOnCompilationFailure`.

File: src/hotspot/share/compiler/compilerOracle.cpp (hunk 3 above)
Comment (shipilev) [reply src/hotspot/share/compiler/compilerOracle.cpp:426]: Moved the `guarantee` closer to `check_predicate`.

**Comment by shipilev** [comment]
There is one unrelated x86_32 test failure, otherwise tests are clean.

/integrate

**Comment by vnkozlov** [comment]
> > Do you need to update blackhole tests to add -XX:+UnlockDiagnosticVMOptions?
> 
> They already have it, because they use `-XX:+AbortVMOnCompilationFailure`.

Good.

**Pull request closed without merging**